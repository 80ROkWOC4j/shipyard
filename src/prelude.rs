pub use crate::get::GetComponent;
pub use crate::iter::{iterators, IntoIter, Shiperator};
pub use crate::not::Not;
pub use crate::remove::Remove;
pub use crate::run::System;
#[doc(hidden)]
pub use crate::run::SystemData;
pub use crate::sparse_set::{sort, sort::Sortable, View, ViewMut};
pub use crate::storage::{AllStorages, Entities, EntitiesMut, Key};
pub use crate::world::World;
pub use crate::Unique;
#[doc(hidden)]
#[cfg(feature = "proc")]
pub use shipyard_proc::system;

#[cfg(feature = "parallel")]
pub use crate::ThreadPool;
