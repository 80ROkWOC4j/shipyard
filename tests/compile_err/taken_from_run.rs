use shipyard::prelude::*;

fn taken_from_run() {
    static mut WORLD: Option<World> = None;

    unsafe { WORLD = Some(World::new::<(usize,)>()) };

    fn test() -> shipyard::internal::ViewMut<'static, usize> {
        let mut result = None;
        unsafe { WORLD.as_ref().unwrap() }.run::<&mut usize, _, _>(|usizes| result = Some(usizes));
        result.unwrap()
    }

    let _view = test();
    test();
}

fn main() {}
