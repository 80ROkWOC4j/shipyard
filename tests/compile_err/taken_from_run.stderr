error[E0373]: closure may outlive the current function, but it borrows `result`, which is owned by the current function
  --> $DIR/taken_from_run.rs:10:68
   |
10 |         unsafe { WORLD.as_ref().unwrap() }.run::<&mut usize, _, _>(|usizes| result = Some(usizes));
   |                                                                    ^^^^^^^^ ------ `result` is borrowed here
   |                                                                    |
   |                                                                    may outlive borrowed value `result`
   |
note: closure is returned here
  --> $DIR/taken_from_run.rs:11:9
   |
11 |         result.unwrap()
   |         ^^^^^^^^^^^^^^^
help: to force the closure to take ownership of `result` (and any other referenced variables), use the `move` keyword
   |
10 |         unsafe { WORLD.as_ref().unwrap() }.run::<&mut usize, _, _>(move |usizes| result = Some(usizes));
   |                                                                    ^^^^^^^^^^^^^

error[E0505]: cannot move out of `result` because it is borrowed
  --> $DIR/taken_from_run.rs:11:9
   |
10 |         unsafe { WORLD.as_ref().unwrap() }.run::<&mut usize, _, _>(|usizes| result = Some(usizes));
   |                                                                    -------- ------ borrow occurs due to use in closure
   |                                                                    |
   |                                                                    borrow of `result` occurs here
11 |         result.unwrap()
   |         ^^^^^^---------
   |         |
   |         move out of `result` occurs here
   |         returning this value requires that `result` is borrowed for `'static`
